pkg/controller/advanceddeployment/controller.go

```diff
@@ -0,0 +1,155 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// controllerKind contains the schema.GroupVersionKind for this DeploymentController type.
+var controllerKind = apps.SchemeGroupVersion.WithKind("Deployment")
+
+// DeploymentController is responsible for synchronizing Deployment objects stored
+// in the system with actual running replica sets and pods.
+type DeploymentController struct {
+	client        client.Client
+	eventRecorder record.EventRecorder
+	strategy      v1alpha1.AdvancedDeploymentStrategy
+}
+
+func NewSyncer(cli client.Client, rec record.EventRecorder, strategy v1alpha1.AdvancedDeploymentStrategy) *DeploymentController {
+	return &DeploymentController{
+		client:        cli,
+		eventRecorder: rec,
+		strategy:      strategy,
+	}
+}
+
+// getReplicaSetsForDeployment uses ControllerRefManager to reconcile
+// ControllerRef by adopting and orphaning.
+// It returns the list of ReplicaSets that this Deployment should manage.
+func (dc *DeploymentController) getReplicaSetsForDeployment(d *apps.Deployment) ([]*apps.ReplicaSet, error) {
+	// List all ReplicaSets to find those we own but that no longer match our
+	// selector. They will be orphaned by ClaimReplicaSets().
+	deploymentSelector, err := metav1.LabelSelectorAsSelector(d.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("deployment %s/%s has invalid label selector: %v", d.Namespace, d.Name, err)
+	}
+	rsLister := &apps.ReplicaSetList{}
+	err = dc.client.List(context.TODO(), rsLister, &client.ListOptions{Namespace: d.Namespace, LabelSelector: deploymentSelector}, utilclient.DisableDeepCopy)
+	if err != nil {
+		return nil, err
+	}
+	rsList := make([]*apps.ReplicaSet, 0, len(rsLister.Items))
+	for i := range rsLister.Items {
+		rs := &rsLister.Items[i]
+		if !rs.DeletionTimestamp.IsZero() {
+			continue
+		}
+		if owner := metav1.GetControllerOfNoCopy(rs); owner == nil || owner.UID != d.UID {
+			continue
+		}
+		rsList = append(rsList, rs)
+	}
+	return rsList, nil
+}
+
+// SyncDeployment will sync the deployment with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (dc *DeploymentController) SyncDeployment(deployment *apps.Deployment) (bool, error) {
+	startTime := time.Now()
+	klog.V(4).InfoS("Started syncing deployment", "deployment", klog.KObj(deployment), "startTime", startTime)
+	defer func() {
+		klog.V(4).InfoS("Finished syncing deployment", "deployment", klog.KObj(deployment), "duration", time.Since(startTime))
+	}()
+
+	// Deep-copy otherwise we are mutating our cache.
+	// TODO: Deep-copy only when needed.
+	d := deployment.DeepCopy()
+
+	// List ReplicaSets owned by this Deployment, while reconciling ControllerRef
+	// through adoption/orphaning.
+	rsList, err := dc.getReplicaSetsForDeployment(d)
+	if err != nil {
+		return false, err
+	}
+
+	if dc.strategy.Paused {
```

Comment:
If it's convenient, hard work comments to identify the core logic of the advanced deployment modified places.